p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,465 @@
+	}
+
+	// do offer-answer exchange
+	offer, err := pc.CreateOffer(nil)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not create offer: %w", err)
+	}
+
+	err = pc.SetLocalDescription(offer)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not set local description: %w", err)
+	}
+
+	answerSdpString := renderServerSdp(raddr, ufrag, remoteMultihash)
+
+	answer := webrtc.SessionDescription{SDP: answerSdpString, Type: webrtc.SDPTypeAnswer}
+	err = pc.SetRemoteDescription(answer)
+	if err != nil {
+		return pc, nil, fmt.Errorf("could not set remote description: %w", err)
+	}
+
+	// await peerconnection opening
+	select {
+	case err := <-errC:
+		if err != nil {
+			return pc, nil, err
+		}
+	case <-ctx.Done():
+		return pc, nil, fmt.Errorf("peerconnection opening timed out")
+	}
+
+	detached, err := getDetachedChannel(ctx, rawHandshakeChannel)
+	if err != nil {
+		return pc, nil, err
+	}
+	// set the local address from the candidate pair
+	cp, err := rawHandshakeChannel.Transport().Transport().ICETransport().GetSelectedCandidatePair()
+	if cp == nil || err != nil {
+		return pc, nil, fmt.Errorf("ice connection did not have selected candidate pair")
+	}
+	laddr := &net.UDPAddr{IP: net.ParseIP(cp.Local.Address), Port: int(cp.Local.Port)}
+
+	channel := newStream(nil, rawHandshakeChannel, detached, laddr, raddr)
+	// the local address of the selected candidate pair should be the
+	// local address for the connection, since different datachannels
+	// are multiplexed over the same SCTP connection
+	localAddr, err := manet.FromNetAddr(channel.LocalAddr())
+	if err != nil {
+		return pc, nil, err
+	}
+
+	// we can only know the remote public key after the noise handshake,
+	// but need to set up the callbacks on the peerconnection
+	conn, err := newConnection(
+		network.DirOutbound,
+		pc,
+		t,
+		scope,
+		t.localPeerId,
+		t.privKey,
+		localAddr,
+		p,
+		nil,
+		remoteMultiaddr,
+	)
+	if err != nil {
+		return pc, nil, err
+	}
+
+	secConn, err := t.noiseHandshake(ctx, pc, channel, p, remoteHashFunction, false)
+	if err != nil {
+		return pc, conn, err
+	}
+	conn.setRemotePublicKey(secConn.RemotePublicKey())
+	return pc, conn, err
+}
+
+func (t *WebRTCTransport) getCertificateFingerprint() (webrtc.DTLSFingerprint, error) {
+	fps, err := t.webrtcConfig.Certificates[0].GetFingerprints()
+	if err != nil {
+		return webrtc.DTLSFingerprint{}, err
+	}
+	return fps[0], nil
+}
+
+func (t *WebRTCTransport) generateNoisePrologue(pc *webrtc.PeerConnection, hash crypto.Hash, inbound bool) ([]byte, error) {
+	raw := pc.SCTP().Transport().GetRemoteCertificate()
+	cert, err := x509.ParseCertificate(raw)
+	if err != nil {
+		return nil, err
+	}
+
+	localFp, err := t.getCertificateFingerprint()
+	if err != nil {
+		return nil, err
+	}
+
+	remoteFp, err := fingerprint.Fingerprint(cert, hash)
+	if err != nil {
+		return nil, err
+	}
+	remoteFp = replaceAll(strings.ToLower(remoteFp), byte(':'))
```

Comment:
TODO[Question]: is the `:` in the string really needed in the first place, seems a human-readable thing
that for machine processing is making things just less efficient / straightforward?!

I get that this is perhaps a spec thing, but still, find it odd...

Reply:
It is needed in the fingerprint spec. From the spec https://www.rfc-editor.org/rfc/rfc4572#section-5

> A fingerprint is represented in SDP as an attribute (an 'a' line).
   It consists of the name of the hash function used, followed by the
   hash value itself.  The hash value is represented as a sequence of
   uppercase hexadecimal bytes, separated by colons.  The number of
   bytes is defined by the hash function.  (This is the syntax used by
   openssl and by the browsers' certificate managers.  It is different
   from the syntax used to represent hash values in, e.g., HTTP digest
   authentication [[18](https://www.rfc-editor.org/rfc/rfc4572#ref-18)], which uses unseparated lowercase hexadecimal
   bytes.  It was felt that consistency with other applications of
   fingerprints was more important.)

Reply:
fixed in https://github.com/libp2p/go-libp2p/pull/1999/commits/f76dfc499857bffe287bedd931eeba6c652a4825 (first commit of new PR https://github.com/libp2p/go-libp2p/pull/1999)